### PR TITLE
dai: nxp: sai: Add support for all synchronization modes

### PR DESCRIPTION
**Pull request by LaurentiuM1234**
Previously, we only supported RX in SYNC with TX. Since the SAI can also support ASYNC RX with ASYNC TX and ASYNC RX with SYNC TX, this patch series adds supports for these scenarios as well.

Also, due to the i.MX93 errata `ERR051421`, we need to add support for the case in which one of the directions is in SYNC with the other while the BYP bit is set for the ASYNC direction.

Below you may find some waveforms proving that BCLK uses the ASYNC direction's divided BCLK instead of the bypassed one when BYP is toggled.

DIV is 0 => RX_BCLK is half of MCLK (should be equal to MCLK or TX_BCLK)
Screenshot from 2023-12-19 13-07-32

(BCLK is half of MCLK's frequency, which is wrong. See commit description and comments from the code for details)

DIV is 1 => RX_BCLK is MCLK / 4 (should be equal to MCLK or TX_BCLK)
Screenshot from 2023-12-19 13-09-51

Note: MCLK is approximately 12.5Mhz in both scenarios.

**EDIT**: Also attaching some register dumps for:

1) First scenario
Screenshot from 2023-12-19 11-33-30

2) Second scenario
Screenshot from 2023-12-19 11-37-39

### CODE REVIEW

File: drivers/dai/nxp/sai/sai.c [hunk 1]
```diff
@@ -353,6 +353,68 @@ static int sai_config_set(const struct device *dev,
 	 * different than the hardware rate (a.k.a this one).
 	 */
 	data->cfg.rate = bespoke->fsync_rate;
+
+#ifdef CONFIG_SAI_IMX93_ERRATA_051421
+	/* a few notes about the following piece of code:
+	 *
```
Comment (dbaluta) [inline drivers/dai/nxp/sai/sai.c:359]: This looks very good. What do you think about adding a function in order to improve readability of sai_config_set?

e.g 
```
#ifdef CONFIG_SAI_IMX93_ERRATA_051421

void sai_config_set_errata_05152(struct *device...)

{ 
//code for errata
}

#else

void void sai_config_set_errata_05152(struct *device...)
{
///
}

#endif

```

File: drivers/dai/nxp/sai/sai.c [hunk 2]
```diff
@@ -528,14 +555,89 @@ static int sai_trigger_stop(const struct device *dev,
...
+		}
+	}
+}
+
 static int sai_trigger_start(const struct device *dev,
 			     enum dai_dir dir)
 {
 	struct sai_data *data;
+	const struct sai_config *cfg;
 	uint32_t old_state;
```
[the remark this answers is not in the document]
Comment (LaurentiuM1234) [reply drivers/dai/nxp/sai/sai.c:698]: Do you mean something like:

Commit 1: introduce DTS properties and assign them to the `tx_config`/`rx_config` variables in `sai_config_set()`
Commit 2: introduce additional logic for the various sync modes

The issue with this is that if you stop at commit 1 and don't take commit 2 you'll end up with something broken (i.e: you can only use RX sync with TX even if your properties allow you to also configure TX sync with RX and TX, RX async)

[the remark this answers is not in the document]
Comment (dbaluta) [reply drivers/dai/nxp/sai/sai.c:698]: Exactly, with the additional quirk that in sai_config_set() will not read tx_config/rx_config properties from dts but will set the default values.

And then in the second patch we will actually use the values from dts.

But if you find that this still breaks the series we can drop the idea.

[the remark this answers is not in the document]
Comment (LaurentiuM1234) [reply drivers/dai/nxp/sai/sai.c:698]: Ah, got it. Will give it a shot. Should work fine.

**Comment by LaurentiuM1234** [comment]
**V2 changes**:

1. Initial commit was split into 2: 1 introducing `rx_sync_mode` and `tx_sync_mode` properties and setting their associated fields in `sai_config`, 2nd introducing the additional logic.
2. Errata logic has been moved to a separate function alongside the wall of text explaining the errata. Should be more readable now.
3. Wall of text from inside `sai_rx_tx_sw_reset()` has been moved to the beginning of the function. Should improve readability as well.

**Review by dbaluta — DISMISSED** [review]

**Comment by dbaluta** [comment]
Looks good to me. Thanks @LaurentiuM1234 . Ready to merge.

File: drivers/dai/nxp/sai/sai.c [hunk 3]
```diff
@@ -383,8 +381,33 @@ static int sai_config_set(const struct device *dev,
...
+	 * is not disabled before the SYNC direction, we can force
+	 * the disablement of the given direction.
+	 */
+	sai_tx_rx_force_disable(dir, data->regmap);
+
+	/* please note the difference between the transmitter/receiver's
+	 * hardware states and their software states. The software
+	 * states can be obtained by reading data->tx/rx_enabled, while
+	 * the hardware states can be obtained by reading TCSR/RCSR. The
+	 * hadrware state can actually differ from the software state.
```
Comment (iuliana-prodan) [inline drivers/dai/nxp/sai/sai.c:471]: s/hadrware /hardware

File: drivers/dai/nxp/sai/sai.c [hunk 4]
```diff
@@ -528,14 +553,77 @@ static int sai_trigger_stop(const struct device *dev,
...
+ *	SYNC and ASYNC for each direction. As such, there's 4 possible combinations
+ *	that need to be covered here:
+ *		a) TX ASYNC, RX ASYNC
+ *		b) TX SYNC, RX ASYNC
+ *		c) TX ASYNC, RX SYNC
+ *		d) TX SYNC, RX SYNC
+ *
+ *	Combination d) is not valid and is covered by a BUILD_ASSERT(). As such, there's 3 valid
+ *	combintations that need to be supported. Since the main branch of the IF statement covers
+ *	combination a), there's only combinatons b) and c) to be covered here.
```
Comment (iuliana-prodan) [inline drivers/dai/nxp/sai/sai.c:640]: s/ combinatons/combinations

File: drivers/dai/nxp/sai/sai.c [hunk 5]
```diff
@@ -142,6 +142,73 @@ static const struct dai_properties
...
+ *	direction will have the BYP bit toggled.
+ *
+ *	IMPORTANT1: in the above diagram and information, RX is SYNC
+ *	with TX. The same applies if RX is SYNC with TX. Also, this
+ *	applies to i.MX93. For other SoCs, things may be different
+ *	so use this information with caution.
+ *
+ *	IMPORTANT2: for this to work, you also need to enable the
+ *	pad's input path. For i.MX93, this can be achieved by setting
+ *	the pad's SION bit.
```
Comment (iuliana-prodan) [inline drivers/dai/nxp/sai/sai.c:191]: As it is now, this works fine on i.MX93 I suppose, right?
The SION bit is set somewhere else? Where?

**Review by iuliana-prodan — COMMENTED** [review]
LGTM.. just some typos

File: drivers/dai/nxp/sai/sai.c (hunk 5 above)
Comment (LaurentiuM1234) [reply drivers/dai/nxp/sai/sai.c:191]: Yep, works fine.
As for the SION bit: it's set via an IOMUXC register (e.g: `SW_MUX_CTL_PAD_GPIO_IO18`). We don't need to set it via this driver as it's already handled by Linux in our setup. One thing to watch out for here is the fact that the `cs42448` DTS for 93 doesn't set this bit so one would have to modify it to support the SYNC-ASYNC scenario (this is why the default configuration is ASNYC-ASYNC).

**Comment by LaurentiuM1234** [comment]
**V3 changes**

- Fixed typos pointed out by @iuliana-prodan. No functional change.

**Review by dbaluta — APPROVED** [review]

**Review by iuliana-prodan — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
(nxp specific, switching assignee)

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**